Curve torsion measurement

cpinter opened the pull request.

lassoan commented on Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.cxx:
```diff
@@ -314,6 +326,176 @@ bool vtkCurveMeasurementsCalculator::CalculatePolyDataCurvature(vtkPolyData* pol
+  if (!this->CurveIsClosed)
+    {
+    // The torsion for the last cell by definition is 0.0 for open curves
+    torsionArray->InsertValue(linePoints->GetId(numberOfPoints-1), 0.0);
```
We should copy the previous distortion value, because that is a better estimation than 0.

lassoan commented on Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.cxx:
```diff
@@ -314,6 +326,176 @@ bool vtkCurveMeasurementsCalculator::CalculatePolyDataCurvature(vtkPolyData* pol
+    torsionArray->SetComponent(linePoints->GetId(0), 0,
+      torsionArray->GetValue(linePoints->GetId(1)));
+    torsionArray->InsertValue(linePoints->GetId(numberOfPoints-1),
+      torsionArray->GetValue(linePoints->GetId(numberOfPoints-2)));
```
Do we handle the cases correctly when we only have 0, 1, or 2 points?

cpinter replied:
I added a check to require at least three points.

lassoan commented on Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.cxx:
```diff
@@ -314,6 +326,176 @@ bool vtkCurveMeasurementsCalculator::CalculatePolyDataCurvature(vtkPolyData* pol
+    vtkErrorMacro("No measurements found");
+    return false;
+    }
+  vtkMRMLMeasurement* firstMeasurement = vtkMRMLMeasurement::SafeDownCast(this->Measurements->GetItemAsObject(0));
```
Shouldn't we look for torsion measurements specifically instead of just the first measurement? If there are no torsion measurements enabled then we may skip the torsion computation completely.

Also, if we need the markup node anyway then it may be cleaner to set the markup node (instead of the measurements object) as input in vtkCurveMeasurementsCalculator.

cpinter replied:
The first measurement at this specific line was used just for getting the node. However, it's indeed cleaner to set/get the MRML node instead, I changed the code accordingly.

lassoan commented on Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.cxx:
```diff
@@ -314,6 +326,176 @@ bool vtkCurveMeasurementsCalculator::CalculatePolyDataCurvature(vtkPolyData* pol
+    normBinormal[1] = binormal[1] / binormalNorm;
+    normBinormal[2] = binormal[2] / binormalNorm;
+
+    // Local torsion
```
could you add a reference to where the computation comes from?

cpinter replied:
Done

lassoan commented on Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.h:
```diff
@@ -30,7 +30,10 @@
 class vtkCallbackCommand;
 
-/// Filter that generates curves between points of an input polydata
+/// Filter that calculates per-point measurements for markups curves.
```
Specify if "per point" is per curve point or per control point.

cpinter replied:
OK. It's per curve point

lassoan commented on Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.h:
```diff
@@ -96,12 +115,16 @@ class VTK_SLICER_MARKUPS_MODULE_MRML_EXPORT vtkCurveMeasurementsCalculator : pub
   vtkCollection* ObservedControlPointArrays;
 
-  std::string CurvatureUnits;
+  std::string CurvatureUnits{"mm-1"};
```
It would be nice to set this based on the length unit in the scene.

cpinter replied:
See `vtkMRMLMarkupsCurveNode::UpdateMeasurementsInternal`. I added torsion unit setting now.

lassoan commented on Modules/Loadable/Markups/MRML/vtkMRMLMarkupsCurveNode.cxx:
```diff
@@ -1410,6 +1437,59 @@ void vtkMRMLMarkupsCurveNode::OnCurvatureMeasurementModified(
+  if (self->CurveMeasurementsCalculator->GetCalculateTorsion() && measurement->GetValueDefined())
+    {
+    // measurement was already on, nothing to do
+    return;
```
How do we know that recomputation is not needed? Shouldn't we check some MTime values?

cpinter replied:
This function handles only enabling/disabling the specific measurements. We discussed it with @lassoan and this part is OK (except that there was a missing call so good that this part has been investigated)

lassoan reviewed (commented):
Thank you, it looks good overall, just added a few comments, mainly just to clarify/document how it all works.

cpinter commented:
Thanks for the review @lassoan, I'll address the comments soon.

cpinter commented on Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.cxx:
```diff
@@ -158,6 +158,18 @@ int vtkCurveMeasurementsCalculator::RequestData(
+    {
+    if (!this->CalculatePolyDataTorsion(outputPolyData))
+      {
+      return 0;
```
Yes, makes a lot of sense. I don't return any more immediately.

cpinter commented:
I addressed all the comments in a separate commit and rebased on the current main. If the commit looks good I think the best would be squashing the last two together and integrating it as two commits. Thanks again.

cpinter commented:
@lassoan Can you please integrate if seems alright? Then the SlicerHeart people can try it in their module. Thank you

lassoan commented on Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.cxx:
```diff
@@ -314,10 +329,180 @@ bool vtkCurveMeasurementsCalculator::CalculatePolyDataCurvature(vtkPolyData* pol
+  // Initialize torsion array
+  vtkNew<vtkDoubleArray> torsionArray;
+  torsionArray->Initialize();
+  torsionArray->SetName("Torsion");
```
This string should be accessible on the vtkCurveMeasurementsCalculator. Preferably it should be possible to set the array name, but at least it there should be a get method for getting the name.

lassoan commented on Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.cxx:
```diff
@@ -314,10 +329,180 @@ bool vtkCurveMeasurementsCalculator::CalculatePolyDataCurvature(vtkPolyData* pol
+    }
+
+  // Initialize torsion array
+  vtkNew<vtkDoubleArray> torsionArray;
```
If it does not make things very complicated, then try to reuse the existing array instead of always creating a new one.

cpinter replied:
I think I tried but it caused some issues. Let me try again.

lassoan commented on Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.cxx:
```diff
@@ -314,10 +329,180 @@ bool vtkCurveMeasurementsCalculator::CalculatePolyDataCurvature(vtkPolyData* pol
+  tangent = tangents->GetTuple3(linePoints->GetId(numberOfPoints-1));
+  currentLength = sqrt( tangent[0]*tangent[0] + tangent[1]*tangent[1] + tangent[2]*tangent[2] );
+  length += currentLength;
+  meanTorsion = meanTorsion / length;
```
Length may be 0 (if points coincide), make sure that it does not cause major trouble.

lassoan commented on Modules/Loadable/Markups/MRML/vtkCurveMeasurementsCalculator.h:
```diff
@@ -66,11 +72,26 @@ class VTK_SLICER_MARKUPS_MODULE_MRML_EXPORT vtkCurveMeasurementsCalculator : pub
+  //@}
+
+  /// Get name of average torsion measurement
+  static const char* GetAverageTorsionName() { return "torsion avg"; };
```
We usually use `mean` instead of average, as it is shorter and maybe a bit more exact. We already use `mean` here for curvature, so let's rename these `GetMeanTorsionName` and `torsion mean`.

cpinter replied:
Ah yes, this was from the algorithm of the first approach that I implemented. I'll rename it.

lassoan reviewed (commented):
It looks good, just added a few trivial comments.

cpinter commented:
Thanks for the comments! I amended the commit resolving each of them.

cpinter commented:
@lassoan Could you please check? Thanks

lassoan commented:
@cpinter Please have a look at this new test failure, probably related, hopefully easy to fix: https://slicer.cdash.org/test/23944053

cpinter commented:
Will do soon!
